module/profiler/profiler.go

```diff
@@ -71,28 +78,52 @@ func (p *AutoProfiler) Enabled() bool {
 	return p.enabled.Load()
 }
 
-func (p *AutoProfiler) Ready() <-chan struct{} {
-	delay := time.Duration(float64(p.interval) * rand.Float64())
-	p.unit.LaunchPeriodically(p.start, p.interval, delay)
+// TriggerRun manually triggers a profile run if one is not already running.
+func (p *AutoProfiler) TriggerRun(d time.Duration) error {
+	select {
+	case p.trigger <- d:
+		return nil
+	default:
+		return errors.New("profiling is already in progress")
+	}
+}
+
+func (p *AutoProfiler) runForever() {
+	jitter := time.Duration(rand.Int63n(int64(p.interval)))
+	t := time.NewTicker(p.interval + jitter)
+	defer t.Stop()
 
-	if !p.Enabled() {
+	for {
+		select {
+		case <-t.C:
+			if p.Enabled() {
+				p.runOnce()
+			}
+		case d := <-p.trigger:
+			oldDuration := p.duration
```

Comment:
nit: maybe change `type profileFunc func(io.Writer) error` to `type profileFunc func(io.Writer, duration) error`, and pass the duration into runOnce as an argument.